feat(client): ts-migrate full /client/utils directory

samarmohan opened the pull request:
Checklist:

<!-- Please follow this checklist and put an x in each of the boxes, like this: [x]. It will ensure that our team takes your pull request seriously. -->

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Everything in the `client/utils` folder was converted!

<!-- Feel free to add any additional description of changes below this line -->
Refers #42256.

samarmohan committed:
Rename /client/utils/*.js files to *.tsx

samarmohan committed:
fix ts problems in /client/utils/*.tsx

ShaunSHamilton reviewed (changes requested):
Hey there,

Thank you, for taking the time to contribute. Could you please change all files that do not use JSX from `.tsx` to `.ts`?

samarmohan committed:
.tsx -> .ts for files that do not use JSX

samarmohan commented:
> Hey there,
> 
> Thank you, for taking the time to contribute. Could you please change all files that do not use JSX from `.tsx` to `.ts`?

Done

samarmohan commented:
@gikf when will the "merge-party" be held?

gikf commented:
> @gikf when will the "merge-party" be held?

It wasn't decided yet, per https://github.com/freeCodeCamp/freeCodeCamp/issues/42256#issuecomment-851673768. Details likely will be announced in that issue.

samarmohan committed:
kebab-case

samarmohan committed:
convert /client/utils/gatsby/*.js to typescript

samarmohan committed:
fix file names

ShaunSHamilton reviewed (changes requested):
Hey again,

As the bot says above, no non-english files should be changed, as we have an automated workflow to rebase them with any changes to English ones, and then they get translated. Please revert the changes to the `i18n` docs.

raisedadead reviewed (commented):
Hey there,

Thank you for taking the time to contribute to the TypeScript migration. 

The main issue has been updated a lot over the last few days, and so we suggest you re-read it and update this pull request to match if you have the time.

The main goals for a PR are:
- [ ] Changing the filename from `PascalCase` to `kebab-case` and ensuring that the imports in code are all still valid.
- [ ] Changing the file extension to .tsx (if a file has React JSX syntax)
- [ ] Fixing as many TypeScript and Eslint errors as you can while refactoring the code.

If you need any assistance, please reach out in our [contributing chat room](https://chat.freecodecamp.org/channel/contributors).

ShaunSHamilton commented on client/utils/build-challenges.ts:
```diff
@@ -1,19 +1,24 @@
   challengesDir,
   getChallengesDirForLang
-} = require('../../curriculum/getChallenges');
-const envData = require('../../config/env.json');
```
@ojeytonwilliams Was there not a reason this was being required instead of imported?

samarmohan replied:
Sorry, I’ll fix that

ShaunSHamilton replied:
@samarmohan There may not be a need to fix this. I am getting clarification from Oliver, whether or not this change could break things.

ojeytonwilliams replied:
No reason I can think of. It won't fail quietly, so if the build works at all with this change it should be fine.

ShaunSHamilton commented on client/utils/challenge-types.ts:
```diff
@@ -1,24 +1,19 @@
 const html = 0;
 const js = 1;
-const backend = 2;
+export const backend = 2;
```
I do not understand why these changes are here.

samarmohan replied:
They were being exported somewhere else, I changed it from a commonjs to an ESM export.

ShaunSHamilton replied:
This is being exported on line 19, already.

samarmohan replied:
I can't re-export a const so instead of doing it around line 19 I did it where it was initialized.

ShaunSHamilton reviewed (changes requested):
Hello there. Thank you, for taking the time to contribute.

Some of these changes should not be here, including the changes to any non-english files.

ShaunSHamilton committed:
chore(client): add package types for migration (#42623)

samarmohan committed:
fix stuff

ojeytonwilliams reviewed (changes requested):
Hey @samarmohan, could you remove the i18n files, please?  

```
docs/i18n/chinese/how-to-work-on-coding-challenges.md
docs/i18n/espanol/how-to-work-on-coding-challenges.md
docs/i18n/portuguese/how-to-work-on-coding-challenges.md
```

Crowdin will create PRs for these files once they've gone through the translation workflow.

Mvrs committed:
feat(client): ts-migrate client/src/components/helpers/** (#42593)

* refactor(client): convert toggle-button to TypeScript

* chore: rename Space to tsx

* refactor(client): convert space to TypeScript

* chore: rename SlimWidthRow to tsx

* refactor(client): slim-width-row to TypeScript

* chore: rename SkeletonSprite to ts

* fix: fixed typos and resolved paths

* chore: resolve path inconsistencies

* refactor(client): skelton-sprite to TypeScript

* chore: rename loader.test to tsx

* chore: add types for react-spinkit

* refactor(client): loader to TypeScript

* refactor(client): link to TypeScript

* refactor(client): image-loader to TypeScript

* refactor(client): full-width-row to TypeScript

* refactor(client): current-challenge-link to TypeScript

* refactor(client): button to TypeScript

* refactor(client): border-color-picker to TypeScript

* refactor(client): avatar-renderer to TypeScript

* chore: changed loadertest(snap) to ts

* chore: optional types added and cleaned files

* fix: args are now optional

* push small updates for Spacer component merge

* update snapshot

* remove type defs from deps

* Revert "remove type defs from deps"

This reverts commit 9f58bf35549d12512bb415dbf57891b7753f3c7f.

* correctly remove client type deps

* final push to remove from deps

Co-authored-by: Shaun Hamilton <shauhami020@gmail.com>

thecodingaviator committed:
feat(client): ts-migrate /client/src/html.js (#42468)

* rename

* migrate

Co-authored-by: Parth Parth <thecodingaviator@users.noreply.github.com>

thecodingaviator committed:
feat(client): ts-migrate /client/src/utils/fixtures/curriculum-helpers-* (#42469)

Co-authored-by: Parth Parth <thecodingaviator@users.noreply.github.com>

thecodingaviator committed:
feat(client): migrate icons to TS (#42453)

* migrate files and update tests

fix: use more memory for gatsby develop (#42433)

feat(client): remove whitespace from calculated values (#42400)

* feat(client?): remove whitespace for calculated values

* feat(client): remove whitespace from styledeclaration

* do not automatically strip

Co-authored-by: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

* fix: include all properties of CSSStyleDeclaration

* fix test for getPropVal

Co-authored-by: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

fix: get showUpcomingChange from env.json (#42440)

fix(client): display legacy certs like current ones (#42038)

* fix: display legacy certs like the current ones

* fix: link projects in legacy certs to project pages

* fix: update tests to changed legacy cert display

* fix: update tests for removed legacy certs forms

* fix: display legacy certs like the current ones

* fix: submit projects for cert on projects pages

* fix: remove legacy certs form submitting handling

* fix: move claiming cert setup before both tests

* fix: remove legacy cert update props and actions

* fix: remove legacy cert updates from api

* fix: correct merge conflict

fix(curriculum): rework Project Euler 98 (#42423)

* fix: rework challenge to use argument in function

* fix: add solution

* fix: use MathJax to improve math notation

fix(curriculum): rework Project Euler 56 (#42364)

* fix: rework challenge to use argument in function

* fix: add solution

* fix: use MathJax to improve look of math notation

fix(curriculum): correct small english typo (#42447)

chore: group together monaco-editor and plugin (#42443)

This should get renovate to create PRs where both are modified.

fix(deps): update dependency algoliasearch to v4.9.2 (#42432)

Co-authored-by: Renovate Bot <bot@renovateapp.com>

migrate files and update tests

Algorithm moved to TSX

first ten files renamed to tsx

first 10 migration complete

first 20 files renamed

migrate some files. rename all

test

index.tsx forced to ignore ts issues

rename and migrate all files

update tests

* remove missed propType declarations

* kebab-caseify

* fi xlinting

Co-authored-by: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>
Co-authored-by: Parth Parth <thecodingaviator@users.noreply.github.com>
Co-authored-by: Shaun Hamilton <shauhami020@gmail.com>

Commit:
feat(client): Migrate OfflineWarning to TS (#42575)

* Rename OfflineWarning to tsx.

* Convert offline-warning to Typescript.

* Avoid using NodeJS.Timeout as type.

thecodingaviator committed:
feat(client): ts-migrate /client/src/analytics (#42457)

* migrate to ts

* fix empty function for linting

* add function return type

Co-authored-by: Parth Parth <thecodingaviator@users.noreply.github.com>

thecodingaviator committed:
feat(client): ts-migrate client/src/assets/images/components (#42462)

* change to tsx

* update imports, complete migration

* kebab-caseify

* make file and component name consistent

* fix function return types and prop types

* update AsSeenIn to keep clear component names

* cypress please start working

Co-authored-by: Parth Parth <thecodingaviator@users.noreply.github.com>

sharmaabhinav committed:
feat(client) ts-migrate AppMountNotifier Component (#42585)

* feat(client) Renamed AppMountNotifier.js to app-mount-notifier.tsx

* feat(client) ts-migrate AppMountNotifier Component

* remove type package from root

Co-authored-by: Shaun Hamilton <shauhami020@gmail.com>

ShaunSHamilton committed:
feat(client): ts-migrate /client-only-routes/ (#42587)

* feat(client): refactor /client-only-routes/ to ts

* ts refactor with no-verify for Spacer

* kebaberise filenames

* fix imports

moT01 committed:
feat(client): ts-migrate challenge templates/show components (#42553)

moT01 committed:
feat(client): ts-migrate challenge templates- challengeTitle + breadCrumbs (#42586)

thecodingaviator committed:
feat(client): Migrate <Link> to ts (#42467)

* rename

* migrate

* fix: remove duplicate

* fix: linting

Co-authored-by: Parth Parth <thecodingaviator@users.noreply.github.com>
Co-authored-by: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

Dripcoding committed:
feat(client): ts-migrate client/src/components/profile (#42378)

* feat: rename Link, Spacer, Profile for typescript

* feaat: migrate Spacer to typescript

* feat: migrate Link to typescript

* feat: migrate Profile to typescript

* feat: migrate Profile test to typescript

* feat: rename Camper.s to Camper.tsx

* feat: migrate Camper to typescript

* feat: rename Certifications

* feat: migrate Certifications to typescript

* feat: rename HeatMap

* feat: migrate HeatMap to typescript

* feat: rename HeatMap.test.

* feat: convert HeatMap.test. to typescript

* feat: make some props optional in ICertificationProps

* feat: rename Portfolio

* feat: migrate Portfolio to typescript

* feat: rename and migrate SocialIcons

* feat: rename TimeLine

* feat: migrate TimeLine to typescript

* feat: rename TimeLine.test.

* feat: migrate TimeLine.test. to typescript

* feat: rename TimelinePagination

* feat: migrate TimelinePagination to typescript

* feat: clean up for typescript migration

Co-authored-by: Shaun Hamilton <shauhami020@gmail.com>

moT01 committed:
feat(client): ts-migrate challenge-templates complete-modal (#42598)

thecodingaviator committed:
feat(client): ts-migrate client/src/components/helpers/form and client/src/components/formHelpers (#42476)

* client(feat): migrate client/src/components/helpers/form

* compete migration

* migrate formHelpers

* migrate BlockSaveButton and BlockSaveWrapper

* finish migration

* kebab-caseify

* Update client/src/components/formHelpers/block-save-button.tsx

Co-authored-by: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

* fix linting errors

* make props optional

Co-authored-by: Parth Parth <thecodingaviator@users.noreply.github.com>
Co-authored-by: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

awu43 committed:
feat(client): ts-migrate client/src/pages (#42445)

* Renamed .js files to .tsx

* Migrated 404 to TS

TODO: Ask about default prop

* Migrated certification to TS

Converted to functional component

* Partially migrated challenges to TS

TODO: Ask about Redirect props

* Installed @types/react-helmet and @types/react-redux

Prep for migrating donate, this caused two new TS errors on 404 and certification

* Migrated donate to TS

TODO: Ask about prop spreading

* Migrated email-sign-up to TS

Converted to functional component and removed unused isSignedIn prop

* Migrated index to TS

Removed unused props

* Migrated learn to TS

* Installed @types/react-instantsearch-dom

Prep for migrating search

* Migrated search to TS

Converted to functional component

* Migrated settings to TS

* Migrated unsubscribed to TS

* Installed @types/validator and @types/lodash-es

Prep for migrating update-email

* Migrated update-email to TS

Converted to functional component

* Migrated user to TS

* Updated effect hook dependencies

Also removed unnecessary comments from 404

* Renamed challenges.test.tsx to .ts

* remove search.tsx, as search.js was removed

* revert: packages

* revert: packages

Co-authored-by: Shaun Hamilton <shauhami020@gmail.com>
Co-authored-by: moT01 <20648924+moT01@users.noreply.github.com>

akshatgarg12 committed:
feat(client): ts-migrate header components (#42287)

thecodingaviator committed:
feat(client): ts-migrate Challenges/components/ (#42582)

Co-authored-by: Parth Parth <thecodingaviator@users.noreply.github.com>

krishna-saurav committed:
feat(client): ts-migrate for files under client/src/component/search folder (#42327)

* typescript migration for files under search folder

* fixing few more eslint errors and warnings

* reverting changes for redux/index.js

* fixing merge conflicts with next

* deleting redux/index.ts

Nirajn2311 committed:
feat(client): ts-migrate Settings component (#42341)

* feat(client): change file format to Typescript

* feat(client): migrate SectionHeader component

* feat(client): migrate ToggleSetting component

* feat(client): migrate Theme component

* feat(client): migrate ResetModal component

* feat(client): migrate DeleteModal component

* feat(client): migrate DangerZone component

* feat(client): migrate Honesty component

* feat(client): migrate Privacy component

* feat(client): migrate Username component

* feat(client): migrate About component

* feat(client): migrate Email component

* feat(client): migrate Internet component

* feat(client): migrate Portfolio component

* feat(client): add required types packages

* feat(client): undo file format change

* feat(client): rename files to kebab-case

* feat(client): rename temp files to kebab-case

* feat(client): Review ts-migration

* Fix imports
* Fix some types
* Remove unncessary comments
* Consistent comment format

* restore mistakenly deleted file

moT01 committed:
revert: i18n docs

samarmohan commented:
@moT01 thank you so so so much!

moT01 commented:
No problem @samarmohan - thanks for contributing to freeCodeCamp 🎉 

I am blocking this for now while we look into a separate issue.

ojeytonwilliams committed:
Fix: assorted next fixes (#42634)

* chore: fix slightly broken package.json

* fix: update children type

* fix: make size optional for Spacer

* fix: correct broken env import

* test: update snapshot and fix typing

* test: fix imports and remove old snapshot

Commit:
chore: pass CI (#42635)

* chore: eslint on icons

* chore: eslint spacer element

* chore: remove more eslint

* fix: existing lint issues

* fix: misnamed import

* fix: client now builds!

* fix: tsc errors

* fix: more linting issues

ShaunSHamilton committed:
update prop-types

ShaunSHamilton committed:
fix show-project-links type linting

raisedadead reviewed (changes requested):
Thanks for the PR. 

I pulled this down, rebase and clean up the commits. 

It looks like this will need a lot of work before we can merge it. With the merge-party over, we are working on making the target branch of this PR stable.

We will not be able to review or merge this PR just yet; you are welcome to try and ensure that the lining and tests are passing if you are interested. If you do, we would also need you to remove all the lining bypasses.

Alternatively, you can leave the PR as is, and we will close it when someone gets around with a replacement PR.

If you have any questions or need help, please reach out to us in the contributors chat. Cheers.
